bin/internal/prepare-environment.sh

```diff
@@ -0,0 +1,197 @@
+  then
+    LAUNCH_COMPONENTS=zss,${LAUNCH_COMPONENTS}
+  fi
+fi
+
+if [[ $LAUNCH_COMPONENTS == *"api-mediation"* ]]
+then
+  # Create the user configurable api-defs
+  STATIC_DEF_CONFIG_DIR=${WORKSPACE_DIR}/api-mediation/api-defs
+  mkdir -p ${STATIC_DEF_CONFIG_DIR}
+fi
+
+# Prepend directory path to all internal components
+INTERNAL_COMPONENTS=""
+for i in $(echo $LAUNCH_COMPONENTS | sed "s/,/ /g")
+do
+  INTERNAL_COMPONENTS=${INTERNAL_COMPONENTS}",${ROOT_DIR}/components/${i}/bin"
+done
+
+# Notes: changed old behavior
+# Old behavior: LAUNCH_COMPONENTS is a list of full path to component lifecycle scripts folder
+#LAUNCH_COMPONENTS=${INTERNAL_COMPONENTS}",${EXTERNAL_COMPONENTS}"
+# New behavior: LAUNCH_COMPONENTS and EXTERNAL_COMPONENTS can be list of component IDs or paths
+#               The other script may use `find_component_directory` function to find the component
+#               root directory.
+LAUNCH_COMPONENTS=${LAUNCH_COMPONENTS}",${EXTERNAL_COMPONENTS}"
+
+###############################
+# Find component root directory
+#
+# This function will find the component in this sequence:
+#   - ${ROOT_DIR}/components/<component-id>
+#   - ${ZWE_EXTENSION_DIR}/<component-id>
+#
+# @param string     component id, or path to component lifecycle scripts
+# Output            component directory will be written to stdout
+find_component_directory() {
+  component_id=$1
+  # find component lifecycle scripts directory
+  component_dir=
+  if [ -d "${component_id}" ]; then
+    component_lifecycle_dir=$component_id
+    if [[ $component_lifecycle_dir == */bin ]]; then
+      # the lifecycle dir ends with /bin, we assume the component root directory is one level up
+      component_dir=$(cd ${component_lifecycle_dir}/../;pwd)
+    else
+      parent_dir=$(cd ${component_lifecycle_dir}/../;pwd)
+      if [ -f "${parent_dir}/manifest.yaml" -o -f "${parent_dir}/manifest.yml" -o -f "${parent_dir}/manifest.json" ]; then
+        # parent directory has manifest file, we assume it's Zowe component manifest and that's the root folder
+        component_dir="${parent_dir}"
+      fi
+    fi
+  else
+    if [ -d "${ROOT_DIR}/components/${component_id}" ]; then
+      # this is a Zowe build-in component
+      component_dir="${ROOT_DIR}/components/${component_id}"
+    elif [ ! -z "${ZWE_EXTENSION_DIR}" ]; then
+      if [ -d "${ZWE_EXTENSION_DIR}/${component_id}" ]; then
+        # this is an extension installed/linked in ZWE_EXTENSION_DIR
+        component_dir="${ZWE_EXTENSION_DIR}/${component_id}"
+      fi
+    fi
+  fi
+
+  echo "$component_dir"
+}
+
+###############################
+# Check if there are errors registered
+#
+# Notes: any error should increase global variable ERRORS_FOUND by 1.
+checkForErrorsFound() {
+  if [[ $ERRORS_FOUND > 0 ]]; then
+    echo "$ERRORS_FOUND errors were found during validatation, please check the message, correct any properties required in ${INSTANCE_DIR}/instance.env and re-launch Zowe"
+    exit $ERRORS_FOUND
```

Comment:
We probably want braces around ERRORS_FOUND. We might also want to canvas opinion if people are happy to exit if any error is found. This was the original behaviour I implemented and wasn't popular, hence removed.